File: reverseNodesInKGroups.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    
    int total(ListNode *start)
    {
        int count=0;
        while(start!=NULL)
        {
            start=start->next;
            count++;
        }
        return count;
    }
    
    ListNode* reverseKGroup(ListNode* head, int k) 
    {
        if(head==NULL)
            return head;
        
      ListNode *prev=NULL,*curr=head,*next_node=NULL;
        
        int len=total(head);
            
        if(k>len)
            return head;
        
        int count=k;
        while(curr!=NULL && count>0)
        {
            next_node=curr->next;
            curr->next=prev;
            prev=curr;
            curr=next_node;
            count--;
        }
        
        if(next_node!=NULL)
        {
            head->next=reverseKGroup(next_node,k); 
        }
        
        return prev;
    }
};

File: assignments/src/Reverse_nodes_in_k_groups.java
// Approach: Handle all the cases. Attach the reversed lists of k size. Join the remaining list
// as it is.
// Time: O(n)
// Space: O(1)

class Reverse_nodes_in_k_groups {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummyHead = new ListNode();
        dummyHead.next = head;

        ListNode curr = head;
        ListNode prev = dummyHead;

        while (curr != null) {
            ListNode start = curr;
            ListNode end = curr;

            // shift end ptr to k-1 places for reversal
            for (int i = 1; i<=k-1 && end != null; i++) {
                end = end.next;
            }

            if (end == null) {
                break;
            }

            // Logic to connect two reversed lists ->
            ListNode endNext = end.next;
            reverse(start, end);
            prev.next = end;
            start.next = endNext;

            prev = start;
            curr = endNext;
        }

        return dummyHead.next;
    }

    private void reverse(ListNode start, ListNode end) {
        ListNode endNext = end.next;
        ListNode prev = null, curr = start;

        while (curr != endNext) {
            ListNode next = curr.next;
            curr.next = prev;

            prev = curr;
            curr = next;
        }
    }
}

class ListNode {
     int val;
     ListNode next;
     ListNode() {}
     ListNode(int val) { this.val = val; }
     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}